spec/controllers/conference_registration_controller_spec.rb

```diff
@@ -0,0 +1,335 @@
+        before do
+          get :show, conference_id: conference_with_open_registration.short_title
+        end
+
+        it 'assigns conference, registration and workshops variables' do
+          expect(assigns(:conference)).to eq conference_with_open_registration
+          expect(assigns(:registration)).to eq @registration
+          expect(assigns(:workshops)).to eq @registration.workshops
+        end
+
+        it 'renders the show template' do
+          expect(response).to render_template('show')
+        end
+      end
+
+      context 'user has purchased a ticket' do
+        before do
+          @ticket = create(:ticket, conference: conference_with_open_registration)
+          @purchased_ticket = create(:ticket_purchase, conference: conference_with_open_registration,
+                                                       user: user,
+                                                       ticket: @ticket)
+          get :show, conference_id: conference_with_open_registration.short_title
+        end
+
+        it 'assigns price of purchased tickets to total_price and purchased tickets to tickets' do
+          expect(assigns(:total_price)).to eq Money.new(10000, 'USD')
+          expect(assigns(:tickets)).to match_array [@purchased_ticket]
+        end
+      end
+
+      context 'user has not purchased any ticket' do
+        before do
+          get :show, conference_id: conference_with_open_registration.short_title
+        end
+
+        it 'assigns 0 dollars to total_price and empty array to tickets variables' do
+          expect(assigns(:total_price)).to eq Money.new(0, 'USD')
+          expect(assigns(:tickets)).to match_array []
+        end
+      end
+    end
+
+    describe 'GET #edit' do
+      before do
+        @registration = create(:registration, conference: conference_with_open_registration, user: user)
+        get :edit, conference_id: conference_with_open_registration.short_title
+      end
+
+      it 'assigns conference and registration variable' do
+        expect(assigns(:conference)).to eq conference_with_open_registration
+        expect(assigns(:registration)).to eq @registration
+      end
+
+      it 'renders the edit template' do
+        expect(response).to render_template('edit')
+      end
+    end
+
+    describe 'POST #create' do
+      context 'successfully registers' do
+        it 'assigns user variable' do
+          post :create, registration: attributes_for(:registration),
+                        conference_id: conference_with_open_registration.short_title
+          expect(assigns(:user)).to eq user
+        end
+
+        it 'creates a new registration' do
+          expect do
+            post :create, registration: attributes_for(:registration),
+                          conference_id: conference_with_open_registration.short_title
+          end.to change{ Registration.count }.by 1
+        end
+      end
+
+      context 'tickets are available' do
```

Comment:
There is one more case:
tickets available, user registers to conference, user buys ticket, user unregisters for conference (ticket is still bought. Note: we don't have a confirmation message for unregistering), user registers again for the conference